core/src/main/java/org/infinispan/stream/impl/local/AbstractLocalCacheStream.java

```diff
@@ -294,9 +295,16 @@ public boolean noneMatch(Predicate<? super R> predicate) {
 
    @Override
    public CloseableIterator<R> iterator() {
-      return Closeables.iterator(getOrCreateStream());
+      // If the stream is null we can assume no intermediate operations and thus we can support remove
+      if (stream == null) {
+         return removableIterator(Closeables.iterator(getOrCreateStream()));
```

Comment:
But the problem is I can only apply the removable iterator if there was no intermediate operations.  After I invoke getOrCreateStream I can't tell if there was any intermediate operations.  Or am I missing your point?